src/query/storage/m3/dynamic_cluster.go

```diff
@@ -88,10 +96,397 @@ func NewDynamicClusterOptions() DynamicClusterOptions {
+	registry, err := cfg.nsInitializer.Init()
+	if err != nil {
+		return err
+	}
+
+	// Get a namespace watch
+	watch, err := registry.Watch()
+	if err != nil {
+		return err
+	}
+
+	// Wait till first namespaces value is received and set the value.
+	d.logger.Info("resolving namespaces with namespace watch", zap.Int("cluster", etcdClusterId))
+	<-watch.C()
+
+	updater := func(namespaces namespace.Map) error {
+		d.updateNamespaces(etcdClusterId, cfg, namespaces)
+		return nil
+	}
+	nsWatch := namespace.NewNamespaceWatch(updater, watch, d.iOpts)
+	if err = nsWatch.Start(); err != nil {
+		return err
+	}
+
+	nsMap := watch.Get()
+	d.updateNamespaces(etcdClusterId, cfg, nsMap)
+
+	d.Lock()
+	d.nsWatches = append(d.nsWatches, nsWatch)
+	d.Unlock()
+
+	return nil
+}
+
+func (d *dynamicCluster) updateNamespaces(
+	etcdClusterId int,
+	clusterCfg DynamicClusterNamespaceConfiguration,
+	newNamespaces namespace.Map,
+) {
+	if newNamespaces == nil {
+		d.logger.Debug("received empty namespace map. ignoring", zap.Int("cluster", etcdClusterId))
+		return
+	}
+
+	d.Lock()
+	defer d.Unlock()
+
+	d.updateNamespacesByEtcdClusterWithLock(etcdClusterId, clusterCfg, newNamespaces)
+	d.updateClusterNamespacesWithLock()
+}
+
+func (d *dynamicCluster) updateNamespacesByEtcdClusterWithLock(
+	etcdClusterId int,
+	clusterCfg DynamicClusterNamespaceConfiguration,
+	newNamespaces namespace.Map,
+) {
+	// TODO(nate): incorporate checking if new namespace is ready once staging state has landed.
+
+	// Check if existing namespaces still exist or need to be updated.
+	if _, ok := d.namespacesByEtcdCluster[etcdClusterId]; !ok {
+		d.namespacesByEtcdCluster[etcdClusterId] = newClusterNamespaceLookup(len(newNamespaces.IDs()))
+	}
+	existing := d.namespacesByEtcdCluster[etcdClusterId]
+	var (
+		sz      = len(newNamespaces.Metadatas())
+		added   = make([]string, 0, sz)
+		updated = make([]string, 0, sz)
+		removed = make([]string, 0, sz)
+	)
+	for nsId, nsMd := range existing.idToMetadata {
+		newNsMd, err := newNamespaces.Get(ident.StringID(nsId))
+		// non-nil error here means namespace is not present (i.e. namespace has been removed)
+		if err != nil {
+			existing.remove(nsId)
+			removed = append(removed, nsId)
+		}
+
+		// Namespace options have been updated; regenerate cluster namespaces.
+		if !nsMd.Equal(newNsMd) {
+			// Replace with new metadata and cluster namespaces.
+			newClusterNamespaces, err := toClusterNamespaces(clusterCfg, newNsMd)
+			if err != nil {
+				// Log error, but don't allow singular failed namespace update to fail all namespace updates.
+				d.logger.Error("failed to update namespace", zap.String("namespace", nsId),
+					zap.Error(err))
+				continue
+			}
+			existing.update(nsId, newNsMd, newClusterNamespaces)
+			updated = append(updated, nsId)
+		}
+	}
+
+	// Check for new namespaces to add.
+	for _, newNsMd := range newNamespaces.Metadatas() {
+		// Namespace has been added.
+		if !existing.exists(newNsMd.ID().String()) {
```

Comment:
nit: flip to `if existing.exists(newNsMd.ID().String()) { continue }` to reduce nesting